nengo_spa/semantic_pointer.py

```diff
@@ -132,6 +138,26 @@ def unitary(self):
             name=self._get_method_name("unitary"),
         )
 
+    def nondegenerate(self):
+        """Make the Semantic Pointer nondegenerate and return as a new object.
+
+        The original object is not modified.
+
+        A degenerate Semantic Pointer is one that cannot be fractionally bound
+        using an arbitrary exponent in an unambiguous manner.
```

Comment:
Wouldn't all unitary SPs *u* be degenerate by that definition because there will always be an exponent *e* such that `u**e == u` (if I'm not mistaken)? And for non-unitary vectors, precision would become a problem if I recall correctly ...

Reply:
It is true that there is an exponent such that `u**e == u`, here I believe unambiguous is meant to mean there is only one output for a possible input. In the HRR case, fractional binding can be thought of as moving along a circle from an origin to the unitary vector taking the shortest direction. If one of the fourier coefficients has a complex angle of +/- pi, then it is on the opposite pole and ambiguous what direction is shortest. The output will depend on which way pi is rounded and could be different on different machines.

Reply:
Yeah, as Brent explains, we found this to be problematic when a complex value is at exactly `-1` (a complex angle of `+/- pi`). Epsion floating point errors can make the difference between its square root being `+i` or `-i`. When this happens, the inverse fft can be inconsistent with itself and produce imaginary values. Let us know if you think of any alternative workarounds.

Reply:
I now have a better understanding of the issue. However, as far as I understand using a real valued Semantic Pointer as input, one half of the Fourier coefficients should be the complex conjugate of the other half and the exponentiation shouldn't change that; except `numpy.fft.fft` is producing coefficients where the complex conjugate stuff doesn't hold if they are close to `-1+0j`. In that case, I would ask why not use `numpy.fft.rfft` to avoid the problem by having a single coefficient as source of truth (so to say). That would still leave some vectors ambiguous (where it also seems important to recognize that this ambiguity comes in degrees depending on the number of coefficients that are close to `-1+0j`), but maybe it does not matter? You couldn't predict in what direction you would rotate on the complex circle, but usually vector are generated randomly anyways, so you wouldn't expect a particular direction. Moreover, this also seems to be an issue for integer binding, if you are a little bit off of `-1+0j` and do multiple bindings.